### DISCUSSION TITLE
Which python executable does the pixi pick?

### DISCUSSION THREAD

**Discussion by yingbo** [prefix-dev/pixi#1931] [Q&A] [answered]
I'm on macOS and new to the pixi. I have system python, pythons from pyenv and anaconda. Under my project folder, if I run python directly, I got an error from pyenv:

```
pyenv: python: command not found

The `python' command exists in these Python versions:
  3.7.16
  3.8.13
  3.9.17
  3.10.9
  3.11.1
  3.12.1
  3.12.2

```

It is OK because usually I either use conda to create a virtualenv, or setup pyenv correctly then use pdm/poetry to create virtualenv . However, when I run pixi, it automatically used python 3.11 and all works. I am just curious, how pixi determines which python to use for the project?  Thanks!

**Comment by baszalmstra** [comment]
Depending on whether you use a pyproject or pixi.toml we read the requirement differently. With a pyproject.toml we read the requires-python field and use that as the requirement for python. For a pixi.toml you have to define a requirement on the python package.

You then probably have something a python requirement that looks something like `>=3.11`. Our solver looks at all available python versions and selects the highest version.

You could also have a requirement like `~=3.11.0` which will select a python 3.11 version.

I hope that answers your question?

**Reply by yingbo** [reply]
Thank you for your quick reply @baszalmstra. I think I understand better now. Just one thing still puzzles me:  I initialized the project using pixi, in pyproject.toml format. The pixi automatically added `requires-python = ">= 3.11"` for me, and create the default env when I added a library to the project. I am curious  which python in my computer the pixi picked at that time. I have python from the macOS, anaconda and pyenv. Since it hardcopies the python, it's hard for me to tell which one it uses. My personal guess is it's from anaconda. 

Thanks!

**Reply by baszalmstra** [reply, marked as answer]
Its installed as part of the environment! If you use the defaults you get it from conda-forge. We hardcoded the 3.11 requirement, it has nothing to do with your system python.

**Reply by yingbo** [reply]
Thanks. Good to know.
